source/Halibut/Transport/Streams/WriteIntoMemoryBufferStream.cs

```diff
@@ -0,0 +1,123 @@
+
+            if (disposing)
+            {
+                memoryStream.Dispose();
+
+                if (onDispose == OnDispose.DisposeInputStream)
+                {
+                    sinkStream.Dispose();
+                }
+            }
+        }
+        
+        public override long Position
+        {
+            get => throw new NotSupportedException();
+            set => throw new NotSupportedException();
+        }
+
+        public override int ReadTimeout
+        {
+            get => throw new NotSupportedException();
+            set => throw new NotSupportedException();
+        }
+
+        public override int WriteTimeout
+        {
+            get => throw new NotSupportedException();
+            set => throw new NotSupportedException();
+        }
+
+        public override void Flush() => throw new NotSupportedException();
+        public override long Seek(long offset, SeekOrigin origin) => throw new NotSupportedException();
+        public override void SetLength(long value) => throw new NotSupportedException();
+        public override int Read(byte[] buffer, int offset, int count) => throw new NotSupportedException();
+
+        public override async Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (!usingSinkStream)
+            {
+                var remainingSpace = writeIntoMemoryLimitBytes - BytesWrittenIntoMemory;
+                if (count <= remainingSpace)
+                {
+                    // We fit completely in memory
+                    await memoryStream.WriteAsync(buffer, offset, count, cancellationToken);
+                    return;
+                }
+
+                // We tried our best, but will no longer fit in memory. Transition over to use the sinkStream.
+                memoryStream.Position = 0;
+                await memoryStream.CopyToAsync(sinkStream, 81920, cancellationToken);
+                usingSinkStream = true;
+            }
+
+            await sinkStream.WriteAsync(buffer, offset, count, cancellationToken);
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            if (!usingSinkStream)
+            {
+                var remainingSpace = writeIntoMemoryLimitBytes - BytesWrittenIntoMemory;
+                if (count <= remainingSpace)
+                {
+                    // We fit completely in memory
+                    memoryStream.Write(buffer, offset, count);
+                    return;
+                }
+
+                // We tried our best, but will no longer fit in memory. Transition over to use the sinkStream.
+                memoryStream.Position = 0;
+                memoryStream.CopyTo(sinkStream);
+                usingSinkStream = true;
+            }
+
+            sinkStream.Write(buffer, offset, count);
+        }
+
+        public async Task WriteAnyUnwrittenDataToSinkStream(CancellationToken cancellationToken)
```

Comment:
Maybe a name like `WriteCacheToUnderlyingStream`, writing to sink stream is confusing since at this point sink/async doesn't mean much.

Reply:
or rename to flush which take's a bool which controls if the underlying stream has flushed called on it.

Reply:
I went with `WriteBufferToUnderlyingStream`. I took your suggestion, but replaced `Cache` with `Buffer`, as that's the nomenclature used in this class.

I did think of using a method like `Flush` initially, but felt it was just a bit too different from what `Flush` actually does.